board/safety/safety_tesla.h

```diff
@@ -0,0 +1,319 @@
+
+int tesla_controls_allowed_last = 0;
+int steer_allowed = 1;
+
+int tesla_brake_prev = 0;
+int tesla_gas_prev = 0;
+int tesla_speed = 0;
+int current_car_time = -1;
+int time_at_last_stalk_pull = -1;
+int eac_status = 0;
+
+int tesla_ignition_started = 0;
+
+static void tesla_rx_hook(CAN_FIFOMailBox_TypeDef *to_push)
+{
+  set_gmlan_digital_output(GMLAN_HIGH);
+  reset_gmlan_switch_timeout(); //we're still in tesla safety mode, reset the timeout counter and make sure our output is enabled
+
+  //int bus_number = (to_push->RDTR >> 4) & 0xFF;
+  uint32_t addr;
+  if (to_push->RIR & 4)
+  {
+    // Extended
+    // Not looked at, but have to be separated
+    // to avoid address collision
+    addr = to_push->RIR >> 3;
+  }
+  else
+  {
+    // Normal
+    addr = to_push->RIR >> 21;
+  }
+
+  // Record the current car time in current_car_time (for use with double-pulling cruise stalk)
+  if (addr == 0x318)
+  {
+    int hour = (to_push->RDLR & 0x1F000000) >> 24;
+    int minute = (to_push->RDHR & 0x3F00) >> 8;
+    int second = (to_push->RDLR & 0x3F0000) >> 16;
+    current_car_time = (hour * 3600) + (minute * 60) + second;
+  }
+
+  if (addr == 0x45)
+  {
+    // 6 bits starting at position 0
+    int lever_position = (to_push->RDLR & 0x3F);
+    if (lever_position == 2)
+    { // pull forward
+      // activate openpilot
+      // TODO: uncomment the if to use double pull to activate
+      //if (current_car_time <= time_at_last_stalk_pull + 1 && current_car_time != -1 && time_at_last_stalk_pull != -1) {
+      controls_allowed = 1;
+      //}
+      time_at_last_stalk_pull = current_car_time;
+    }
+    else if (lever_position == 1)
+    { // push towards the back
+      // deactivate openpilot
+      controls_allowed = 0;
+    }
+  }
+
+  // Detect drive rail on (ignition) (start recording)
+  if (addr == 0x348)
+  {
+    // GTW_status
+    int drive_rail_on = (to_push->RDLR & 0x0001);
+    tesla_ignition_started = drive_rail_on == 1;
+  }
+
+  // exit controls on brake press
+  // DI_torque2::DI_brakePedal 0x118
+  if (addr == 0x118)
+  {
+    // 1 bit at position 16
+    if (((to_push->RDLR & 0x8000)) >> 15 == 1)
+    {
+      //disable break cancel by commenting line below
+      controls_allowed = 0;
+    }
+    //get vehicle speed in m/s. Tesla gives MPH
+    tesla_speed = ((((((to_push->RDLR >> 24) & 0x0F) << 8) + ((to_push->RDLR >> 16) & 0xFF)) * 0.05 - 25) * 1.609 / 3.6);
+    if (tesla_speed < 0)
+    {
+      tesla_speed = 0;
+    }
+  }
+
+  // exit controls on EPAS error
+  // EPAS_sysStatus::EPAS_eacStatus 0x370
+  if (addr == 0x370)
+  {
+    // if EPAS_eacStatus is not 1 or 2, disable control
+    eac_status = ((to_push->RDHR >> 21)) & 0x7;
+    // For human steering override we must not disable controls when eac_status == 0
+    // Additional safety: we could only allow eac_status == 0 when we have human steerign allowed
+    if ((controls_allowed == 1) && (eac_status != 0) && (eac_status != 1) && (eac_status != 2))
```

Comment:
If we don't want to disable on user steer override (let's think about it), then I would remove this check. Does not seem safety relevant.

Reply:
yeah, it would be safety relevant only if we want to disengage on user steer override.
I assume that if EAC is faulted, then it won't respond to steer commands.

Reply:
Yes, as soon as EAC faults, it disables steer commands and the user must take control.  However, there is no feedback to openpilot indicating this, so I'm thinking it would be good for us to set controls_allowed = 0 because then feedback gets back to OP and it can throw an alert on the screen and/or play an alert sound.  What do you think?